File: app/src/main/java/com/mithrilmania/blocktopograph/map/Biome.java
package com.mithrilmania.blocktopograph.map;

import android.util.SparseArray;

import com.mithrilmania.blocktopograph.util.ColorWrapper;


/*
Biome enum for MCPE -- by @mithrilmania
Reference link: https://minecraft.fandom.com/wiki/Biome/ID

--- Please attribute @mithrilmania for generating+updating this enum
 */
public enum Biome {

	OCEAN(0, "Ocean", ColorWrapper.fromRGB(2, 0, 112)),
	PLAINS(1, "Plains", ColorWrapper.fromRGB(140, 176, 96)),
	DESERT(2, "Desert", ColorWrapper.fromRGB(251, 148, 27)),
	EXTREME_HILLS(3, "Mountains", ColorWrapper.fromRGB(93, 99, 93)),
	FOREST(4, "Forest", ColorWrapper.fromRGB(2, 99, 32)),
	TAIGA(5, "Taiga", ColorWrapper.fromRGB(9, 102, 91)),
	SWAMPLAND(6, "Swampland", ColorWrapper.fromRGB(4, 200, 139)),
	RIVER(7, "River", ColorWrapper.fromRGB(1, 1, 255)),
	HELL(8, "Nether Wastes", ColorWrapper.fromRGB(132, 65, 65)),
	THE_END(9, "The End", ColorWrapper.fromRGB(130, 129, 254)),
	FROZEN_OCEAN(10, "Frozen Ocean", ColorWrapper.fromRGB(142, 141, 161)),
	FROZEN_RIVER(11, "Frozen River", ColorWrapper.fromRGB(159, 163, 255)),
	ICE_PLAINS(12, "Ice Plains", ColorWrapper.fromRGB(255, 254, 255)),
	ICE_MOUNTAINS(13, "Ice Mountains", ColorWrapper.fromRGB(162, 157, 157)),
	MUSHROOM_ISLAND(14, "Mushroom Fields", ColorWrapper.fromRGB(254, 1, 255)),
	MUSHROOM_ISLAND_SHORE(15, "Mushroom Fields Shore", ColorWrapper.fromRGB(158, 3, 253)),
	BEACH(16, "Beach", ColorWrapper.fromRGB(250, 223, 85)),
	DESERT_HILLS(17, "Desert Hills", ColorWrapper.fromRGB(212, 94, 15)),
	FOREST_HILLS(18, "Wooded Hills", ColorWrapper.fromRGB(37, 86, 30)),
	TAIGA_HILLS(19, "Taiga Hills", ColorWrapper.fromRGB(25, 54, 49)),
	EXTREME_HILLS_EDGE(20, "Mountain Edge", ColorWrapper.fromRGB(115, 118, 157)),
	JUNGLE(21, "Jungle", ColorWrapper.fromRGB(82, 122, 7)),
	JUNGLE_HILLS(22, "Jungle Hills", ColorWrapper.fromRGB(46, 64, 3)),
	JUNGLE_EDGE(23, "Jungle Edge", ColorWrapper.fromRGB(99, 142, 24)),
	DEEP_OCEAN(24, "Deep Ocean", ColorWrapper.fromRGB(2, 0, 47)),
	STONE_BEACH(25, "Stone Shore", ColorWrapper.fromRGB(162, 164, 132)),
	COLD_BEACH(26, "Snowy Beach", ColorWrapper.fromRGB(250, 238, 193)),
	BIRCH_FOREST(27, "Birch Forest", ColorWrapper.fromRGB(48, 117, 70)),
	BIRCH_FOREST_HILLS(28, "Birch Forest Hills", ColorWrapper.fromRGB(29, 94, 51)),
	ROOFED_FOREST(29, "Dark Forest", ColorWrapper.fromRGB(66, 82, 24)),
	COLD_TAIGA(30, "Snowy Taiga", ColorWrapper.fromRGB(49, 85, 75)),
	COLD_TAIGA_HILLS(31, "Snowy Taiga Hills", ColorWrapper.fromRGB(34, 61, 52)),
	MEGA_TAIGA(32, "Giant Tree Taiga", ColorWrapper.fromRGB(92, 105, 84)),
	MEGA_TAIGA_HILLS(33, "Giant Tree Taiga Hills", ColorWrapper.fromRGB(70, 76, 59)),
	EXTREME_HILLS_PLUS_TREE(34, "Wooded Mountains", ColorWrapper.fromRGB(79, 111, 81)),
	SAVANNA(35, "Savanna", ColorWrapper.fromRGB(192, 180, 94)),
	SAVANNA_PLATEAU(36, "Savanna Plateau", ColorWrapper.fromRGB(168, 157, 98)),
	MESA(37, "Badlands", ColorWrapper.fromRGB(220, 66, 19)),
	MESA_PLATEAU(38, "Badlands Plateau", ColorWrapper.fromRGB(174, 152, 100)),
	MESA_PLATEAU_STONE(39, "Wooded Badlands Plateau", ColorWrapper.fromRGB(202, 139, 98)),

	//fix the colors for these
	WARM_OCEAN(40, "Warm Ocean", ColorWrapper.fromRGB(202, 139, 98)),
	LUKEWARM_OCEAN(41, "Lukewarm Ocean", ColorWrapper.fromRGB(202, 139, 98)),
	COLD_OCEAN(42, "Cold Ocean", ColorWrapper.fromRGB(202, 139, 98)),
	DEEP_WARM_OCEAN(43, "Deep Warm Ocean", ColorWrapper.fromRGB(202, 139, 98)),
	DEEP_LUKEWARM_OCEAN(44, "Deep Lukewarm Ocean", ColorWrapper.fromRGB(202, 139, 98)),
	DEEP_COLD_OCEAN(45, "Deep Cold Ocean", ColorWrapper.fromRGB(202, 139, 98)),
	DEEP_FROZEN_OCEAN(46, "Deep Frozen Ocean", ColorWrapper.fromRGB(202, 139, 98)),
	LEGACY_FROZEN_OCEAN(47, "Legacy Frozen Ocean", ColorWrapper.fromRGB(202, 139, 98)),

	//  Not sure are available ( need tester )
//	OCEAN_M(128, "Ocean M", ColorWrapper.fromRGB(81, 79, 195)),

	SUNFLOWER_PLAINS(129, "Sunflower Plains", ColorWrapper.fromRGB(220, 255, 177)),
	DESERT_MUTATED(130, "Desert Lakes", ColorWrapper.fromRGB(255, 230, 101)),
	EXTREME_HILLS_MUTATED(131, "Gravelly Mountains", ColorWrapper.fromRGB(177, 176, 174)),
	FLOWER_FOREST(132, "Flower Forest", ColorWrapper.fromRGB(82, 180, 110)),
	TAIGA_MUTATED(133, "Taiga Mountains", ColorWrapper.fromRGB(90, 182, 171)),
	SWAMPLAND_MUTATED(134, "Swamp Hills", ColorWrapper.fromRGB(87, 255, 255)),

	// Not sure available ( need tester )
//	RIVER_M(135, "River M", ColorWrapper.fromRGB(82, 79, 255)),
//	HELL_M(136, "Hell M", ColorWrapper.fromRGB(255, 80, 83)),
//	SKY_M(137, "Sky M", ColorWrapper.fromRGB(210, 211, 255)),
//	FROZEN_OCEAN_M(138, "Frozen Ocean M", ColorWrapper.fromRGB(226, 224, 241)),
//	FROZEN_RIVER_M(139, "Frozen River M", ColorWrapper.fromRGB(239, 242, 255)),

	ICE_PLAINS_SPIKES(140, "Ice Spikes", ColorWrapper.fromRGB(223, 255, 255)),

	// Not sure available ( need tester )
//	ICE_MOUNTAINS_M(141, "Ice Mountains M", ColorWrapper.fromRGB(237, 237, 238)),
//	MUSHROOM_ISLAND_M(142, "Mushroom Island M", ColorWrapper.fromRGB(255, 82, 255)),
//	MUSHROOM_ISLAND_SHORE_M(143, "Mushroom Island Shore M", ColorWrapper.fromRGB(243, 82, 255)),
//	BEACH_M(144, "Beach M", ColorWrapper.fromRGB(255, 255, 162)),
//	DESERT_HILLS_M(145, "Desert Hills M", ColorWrapper.fromRGB(255, 177, 100)),
//	FOREST_HILLS_M(146, "Forest Hills M", ColorWrapper.fromRGB(113, 167, 109)),
//	TAIGA_HILLS_M(147, "Taiga Hills M", ColorWrapper.fromRGB(103, 135, 134)),
//	EXTREME_HILLS_EDGE_M(148, "Extreme Hills Edge M", ColorWrapper.fromRGB(196, 203, 234)),

	JUNGLE_MUTATED(149, "Modified Jungle", ColorWrapper.fromRGB(160, 203, 92)),
//	JUNGLE_HILLS_M(150, "Jungle Hills M", ColorWrapper.fromRGB(127, 146, 86)),
	JUNGLE_EDGE_MUTATED(151, "Modified Jungle Edge", ColorWrapper.fromRGB(179, 217, 105)),
//	DEEP_OCEAN_M(152, "Deep Ocean M", ColorWrapper.fromRGB(82, 79, 130)),
//	STONE_BEACH_M(153, "Stone Beach M", ColorWrapper.fromRGB(242, 243, 209)),
//	COLD_BEACH_M(154, "Cold Beach M", ColorWrapper.fromRGB(255, 255, 255)),
	BIRCH_FOREST_MUTATED(155, "Tall Birch Forest", ColorWrapper.fromRGB(131, 194, 148)),
	BIRCH_FOREST_HILLS_MUTATED(156, "Tall Birch Hills", ColorWrapper.fromRGB(111, 175, 133)),
	ROOFED_FOREST_MUTATED(157, "Dark Forest Hills", ColorWrapper.fromRGB(143, 158, 109)),
	COLD_TAIGA_MUTATED(158, "Snowy Taiga Mountains", ColorWrapper.fromRGB(132, 163, 156)),
//	COLD_TAIGA_HILLS_M(159, "Cold Taiga Hills M", ColorWrapper.fromRGB(113, 143, 136)),
	REDWOOD_TAIGA_MUTATED(160, "Giant Spruce Taiga", ColorWrapper.fromRGB(168, 180, 164)),
	REDWOOD_TAIGA_HILLS_MUTATED(161, "Giant Spruce Taiga Hills", ColorWrapper.fromRGB(150, 158, 140)),
	EXTREME_HILLS_PLUS_TREE_MUTATED(162, "Gravelly Mountains", ColorWrapper.fromRGB(161, 194, 158)),
	SAVANNA_MUTATED(163, "Shattered Savanna", ColorWrapper.fromRGB(255, 255, 173)),
	SAVANNA_PLATEAU_MUTATED(164, "Shattered Savanna Plateau", ColorWrapper.fromRGB(247, 238, 180)),
	MESA_BRYCE(165, "Eroded Badlands", ColorWrapper.fromRGB(255, 151, 101)),
	MESA_PLATEAU_MUTATED(166, "Modified Badlands Plateau", ColorWrapper.fromRGB(255, 234, 179)),
	MESA_PLATEAU_STONE_MUTATED(167, "Modified Wooded Badlands Plateau", ColorWrapper.fromRGB(255, 220, 184)),
	BAMBOO_JUNGLE(168, "Bamboo Jungle", ColorWrapper.fromRGB(255, 220, 184)),
	BAMBOO_JUNGLE_HILLS(169, "Bamboo Jungle Hills", ColorWrapper.fromRGB(255, 220, 184)),

	// 1.16 BIOME
	SOUL_SAND_VALLEY(178, "Soul Sand Valley", ColorWrapper.fromRGB(66, 113, 114)),
	CRIMSON_FOREST(179, "Crimson Forest", ColorWrapper.fromRGB(141, 30, 40)),
	WARPED_FOREST(180, "Warped Forest", ColorWrapper.fromRGB(22, 126, 134)),
	BASALT_DELTAS(181, "Basalt Deltas", ColorWrapper.fromRGB(75, 69, 71)),

	// 1.17 BIOME ?
	JAGGED_PEAKS(182, "Jagged Peaks", ColorWrapper.fromRGB(0, 0, 0)),
	FROZEN_PEAKS(183, "Frozen Peaks", ColorWrapper.fromRGB(0, 0, 0)),
	SNOWY_SLOPES(184, "Snowy Slopes", ColorWrapper.fromRGB(0, 0, 0)),
	GROVE(185, "Grove", ColorWrapper.fromRGB(0, 0, 0)),
	MEADOW(186, "Meadow", ColorWrapper.fromRGB(0, 0, 0)),
	LUSH_CAVES(187, "Lush Caves", ColorWrapper.fromRGB(0, 0, 0)),
	DRIPSTONE_CAVES(188, "Dripstone Caves", ColorWrapper.fromRGB(0, 0, 0)),
	STONY_PEAKS(189, "Stony Peaks", ColorWrapper.fromRGB(0, 0, 0))
	;

	private static final SparseArray<Biome> biomeMap;

	static {
		biomeMap = new SparseArray<>();
		for (Biome b : Biome.values()) {
			biomeMap.put(b.id, b);
		}
	}

	public final int id;
	public final String name;
	public final ColorWrapper color;

	Biome(int id, String name, ColorWrapper color) {
		this.id = id;
		this.name = name;
		this.color = color;
	}

	public static Biome getBiome(int id) {
		return biomeMap.get(id);
	}


	public String getName() {
		return name;
	}
}

File: src/server/core/chunk/biome.rs
/**
 * Copyright © 2025 NguyenDuck
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
////////////////////////////////////////////////////////////////////////
use std::io::Error;

macro_rules! define_biome_id {
    (
        $(#[$meta:meta])*
        pub enum $name:ident {
            $($variant:ident = $id:literal,)*
        }
    ) => {
        $(#[$meta])*
        pub enum $name {
            $($variant,)*
        }

        impl $name {
            const ALL: &'static [$name] = &[$($name::$variant,)*];

            pub fn iter() -> impl Iterator<Item = &'static $name> {
                Self::ALL.iter()
            }

            pub fn to_i32(&self) -> i32 {
                match self {
                    $($name::$variant => $id,)*
                }
            }

            pub fn from_i32(id: i32) -> Result<Self, Error> {
                match id {
                    $($id => Ok($name::$variant),)*
                    _ => Err(Error::new(std::io::ErrorKind::InvalidData, format!("Unknown biome id: {}", id))),
                }
            }
        }
    };
}

define_biome_id! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
    pub enum BiomeId {
        Ocean = 0,
        Plains = 1,
        Desert = 2,
        ExtremeHills = 3,
        Forest = 4,
        Taiga = 5,
        Swampland = 6,
        River = 7,
        Hell = 8,
        TheEnd = 9,
        LegacyFrozenOcean = 10,
        FrozenRiver = 11,
        IcePlains = 12,
        IceMountains = 13,
        MushroomIsland = 14,
        MushroomIslandShore = 15,
        Beach = 16,
        DesertHills = 17,
        ForestHills = 18,
        TaigaHills = 19,
        ExtremeHillsEdge = 20,
        Jungle = 21,
        JungleHills = 22,
        JungleEdge = 23,
        DeepOcean = 24,
        StoneBeach = 25,
        ColdBeach = 26,
        BirchForest = 27,
        BirchForestHills = 28,
        RoofedForest = 29,
        ColdTaiga = 30,
        ColdTaigaHills = 31,
        MegaTaiga = 32,
        MegaTaigaHills = 33,
        ExtremeHillsPlusTrees = 34,
        Savanna = 35,
        SavannaPlateau = 36,
        Mesa = 37,
        MesaPlateauStone = 38,
        MesaPlateau = 39,
        WarmOcean = 40,
        DeepWarmOcean = 41,
        LukewarmOcean = 42,
        DeepLukewarmOcean = 43,
        ColdOcean = 44,
        DeepColdOcean = 45,
        FrozenOcean = 46,
        DeepFrozenOcean = 47,
        BambooJungle = 48,
        BambooJungleHills = 49,
        SunflowerPlains = 129,
        DesertMutated = 130,
        ExtremeHillsMutated = 131,
        FlowerForest = 132,
        TaigaMutated = 133,
        SwamplandMutated = 134,
        IcePlainsSpikes = 140,
        JungleMutated = 149,
        JungleEdgeMutated = 151,
        BirchForestMutated = 155,
        BirchForestHillsMutated = 156,
        RoofedForestMutated = 157,
        ColdTaigaMutated = 158,
        RedwoodTaigaMutated = 160,
        RedwoodTaigaHillsMutated = 161,
        ExtremeHillsPlusTreesMutated = 162,
        SavannaMutated = 163,
        SavannaPlateauMutated = 164,
        MesaBryce = 165,
        MesaPlateauStoneMutated = 166,
        MesaPlateauMutated = 167,
        SoulsandValley = 178,
        CrimsonForest = 179,
        WarpedForest = 180,
        BasaltDeltas = 181,
        JaggedPeaks = 182,
        FrozenPeaks = 183,
        SnowySlopes = 184,
        Grove = 185,
        Meadow = 186,
        LushCaves = 187,
        DripstoneCaves = 188,
        StonyPeaks = 189,
        DeepDark = 190,
        MangroveSwamp = 191,
        CherryGrove = 192,
        PaleGarden = 193,
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Biome {
    id: BiomeId,
    temperature: f32,
    humidity: f32,
    resource_location: &'static str,
}

impl Biome {
    const fn new(
        id: BiomeId,
        temperature: f32,
        humidity: f32,
        resource_location: &'static str,
    ) -> Self {
        Self {
            id,
            temperature,
            humidity,
            resource_location,
        }
    }

    pub const ALL: &'static [Biome] = &[
        Self::new(BiomeId::BambooJungle, 0.9, 0.9, "bamboo_jungle"),
        Self::new(BiomeId::BambooJungleHills, 0.9, 0.9, "bamboo_jungle_hills"),
        Self::new(BiomeId::BasaltDeltas, 2.0, 0.0, "basalt_deltas"),
        Self::new(BiomeId::Beach, 0.8, 0.4, "beach"),
        Self::new(BiomeId::BirchForest, 0.6, 0.6, "birch_forest"),
        Self::new(BiomeId::BirchForestHills, 0.6, 0.6, "birch_forest_hills"),
        Self::new(
            BiomeId::BirchForestHillsMutated,
            0.7,
            0.8,
            "birch_forest_hills_mutated",
        ),
        Self::new(
            BiomeId::BirchForestMutated,
            0.6,
            0.6,
            "birch_forest_mutated",
        ),
        Self::new(BiomeId::CherryGrove, 0.3, 0.8, "cherry_grove"),
        Self::new(BiomeId::ColdBeach, 0.1, 0.3, "cold_beach"),
        Self::new(BiomeId::ColdOcean, 0.5, 0.5, "cold_ocean"),
        Self::new(BiomeId::ColdTaiga, -0.5, 0.4, "cold_taiga"),
        Self::new(BiomeId::ColdTaigaHills, -0.5, 0.4, "cold_taiga_hills"),
        Self::new(BiomeId::ColdTaigaMutated, -0.5, 0.4, "cold_taiga_mutated"),
        Self::new(BiomeId::CrimsonForest, 2.0, 0.0, "crimson_forest"),
        Self::new(BiomeId::DeepColdOcean, 0.5, 0.5, "deep_cold_ocean"),
        Self::new(BiomeId::DeepDark, 0.8, 0.4, "deep_dark"),
        Self::new(BiomeId::DeepFrozenOcean, 0.0, 0.5, "deep_frozen_ocean"),
        Self::new(BiomeId::DeepLukewarmOcean, 0.5, 0.5, "deep_lukewarm_ocean"),
        Self::new(BiomeId::DeepOcean, 0.5, 0.5, "deep_ocean"),
        Self::new(BiomeId::DeepWarmOcean, 0.5, 0.5, "deep_warm_ocean"),
        Self::new(BiomeId::Desert, 2.0, 0.0, "desert"),
        Self::new(BiomeId::DesertHills, 2.0, 0.0, "desert_hills"),
        Self::new(BiomeId::DesertMutated, 2.0, 0.0, "desert_mutated"),
        Self::new(BiomeId::DripstoneCaves, 0.2, 0.0, "dripstone_caves"),
        Self::new(BiomeId::ExtremeHills, 0.2, 0.3, "extreme_hills"),
        Self::new(BiomeId::ExtremeHillsEdge, 0.2, 0.3, "extreme_hills_edge"),
        Self::new(
            BiomeId::ExtremeHillsMutated,
            0.2,
            0.3,
            "extreme_hills_mutated",
        ),
        Self::new(
            BiomeId::ExtremeHillsPlusTrees,
            0.2,
            0.3,
            "extreme_hills_plus_trees",
        ),
        Self::new(
            BiomeId::ExtremeHillsPlusTreesMutated,
            0.2,
            0.3,
            "extreme_hills_plus_trees_mutated",
        ),
        Self::new(BiomeId::FlowerForest, 0.7, 0.8, "flower_forest"),
        Self::new(BiomeId::Forest, 0.7, 0.8, "forest"),
        Self::new(BiomeId::ForestHills, 0.7, 0.8, "forest_hills"),
        Self::new(BiomeId::FrozenOcean, 0.0, 0.5, "frozen_ocean"),
        Self::new(BiomeId::FrozenPeaks, -0.7, 0.9, "frozen_peaks"),
        Self::new(BiomeId::FrozenRiver, 0.0, 0.5, "frozen_river"),
        Self::new(BiomeId::Grove, -0.2, 0.8, "grove"),
        Self::new(BiomeId::Hell, 2.0, 0.0, "hell"),
        Self::new(BiomeId::IceMountains, 0.0, 0.5, "ice_mountains"),
        Self::new(BiomeId::IcePlains, 0.0, 0.5, "ice_plains"),
        Self::new(BiomeId::IcePlainsSpikes, 0.0, 1.0, "ice_plains_spikes"),
        Self::new(BiomeId::JaggedPeaks, -0.7, 0.9, "jagged_peaks"),
        Self::new(BiomeId::Jungle, 0.9, 0.9, "jungle"),
        Self::new(BiomeId::JungleEdge, 0.9, 0.8, "jungle_edge"),
        Self::new(BiomeId::JungleEdgeMutated, 0.9, 0.8, "jungle_edge_mutated"),
        Self::new(BiomeId::JungleHills, 0.9, 0.9, "jungle_hills"),
        Self::new(BiomeId::JungleMutated, 0.9, 0.9, "jungle_mutated"),
        Self::new(BiomeId::LegacyFrozenOcean, 0.0, 0.5, "legacy_frozen_ocean"),
        Self::new(BiomeId::LukewarmOcean, 0.5, 0.5, "lukewarm_ocean"),
        Self::new(BiomeId::LushCaves, 0.9, 0.0, "lush_caves"),
        Self::new(BiomeId::MangroveSwamp, 0.8, 0.9, "mangrove_swamp"),
        Self::new(BiomeId::Meadow, 0.3, 0.8, "meadow"),
        Self::new(BiomeId::MegaTaiga, 0.3, 0.8, "mega_taiga"),
        Self::new(BiomeId::MegaTaigaHills, 0.3, 0.8, "mega_taiga_hills"),
        Self::new(BiomeId::Mesa, 2.0, 0.0, "mesa"),
        Self::new(BiomeId::MesaBryce, 2.0, 0.0, "mesa_bryce"),
        Self::new(BiomeId::MesaPlateau, 2.0, 0.0, "mesa_plateau"),
        Self::new(
            BiomeId::MesaPlateauMutated,
            2.0,
            0.0,
            "mesa_plateau_mutated",
        ),
        Self::new(BiomeId::MesaPlateauStone, 2.0, 0.0, "mesa_plateau_stone"),
        Self::new(
            BiomeId::MesaPlateauStoneMutated,
            2.0,
            0.0,
            "mesa_plateau_stone_mutated",
        ),
        Self::new(BiomeId::MushroomIsland, 0.9, 1.0, "mushroom_island"),
        Self::new(
            BiomeId::MushroomIslandShore,
            0.9,
            1.0,
            "mushroom_island_shore",
        ),
        Self::new(BiomeId::Ocean, 0.5, 0.5, "ocean"),
        Self::new(BiomeId::PaleGarden, 0.7, 0.8, "pale_garden"),
        Self::new(BiomeId::Plains, 0.8, 0.4, "plains"),
        Self::new(
            BiomeId::RedwoodTaigaHillsMutated,
            0.3,
            0.8,
            "redwood_taiga_hills_mutated",
        ),
        Self::new(
            BiomeId::RedwoodTaigaMutated,
            0.3,
            0.8,
            "redwood_taiga_mutated",
        ),
        Self::new(BiomeId::River, 0.5, 0.5, "river"),
        Self::new(BiomeId::RoofedForest, 0.7, 0.8, "roofed_forest"),
        Self::new(
            BiomeId::RoofedForestMutated,
            0.7,
            0.8,
            "roofed_forest_mutated",
        ),
        Self::new(BiomeId::Savanna, 1.2, 0.0, "savanna"),
        Self::new(BiomeId::SavannaMutated, 2.0, 0.0, "savanna_mutated"),
        Self::new(BiomeId::SavannaPlateau, 1.0, 0.0, "savanna_plateau"),
        Self::new(
            BiomeId::SavannaPlateauMutated,
            1.0,
            0.5,
            "savanna_plateau_mutated",
        ),
        Self::new(BiomeId::SnowySlopes, -0.3, 0.9, "snowy_slopes"),
        Self::new(BiomeId::SoulsandValley, 2.0, 0.0, "soulsand_valley"),
        Self::new(BiomeId::StoneBeach, 0.2, 0.3, "stone_beach"),
        Self::new(BiomeId::StonyPeaks, 1.0, 0.3, "stony_peaks"),
        Self::new(BiomeId::SunflowerPlains, 0.8, 0.4, "sunflower_plains"),
        Self::new(BiomeId::Swampland, 0.8, 0.9, "swampland"),
        Self::new(BiomeId::SwamplandMutated, 0.8, 0.5, "swampland_mutated"),
        Self::new(BiomeId::Taiga, 0.3, 0.8, "taiga"),
        Self::new(BiomeId::TaigaHills, 0.3, 0.8, "taiga_hills"),
        Self::new(BiomeId::TaigaMutated, 0.3, 0.8, "taiga_mutated"),
        Self::new(BiomeId::TheEnd, 0.5, 0.5, "the_end"),
        Self::new(BiomeId::WarmOcean, 0.5, 0.5, "warm_ocean"),
        Self::new(BiomeId::WarpedForest, 2.0, 0.0, "warped_forest"),
    ];

    pub fn from_id(id: i32) -> Result<Self, Error> {
        Self::ALL
            .iter()
            .find(|biome| biome.id.to_i32() == id)
            .ok_or(Error::new(
                std::io::ErrorKind::InvalidData,
                format!("Unknown biome id: {}", id),
            ))
            .cloned()
    }

    pub fn id(&self) -> BiomeId {
        self.id
    }

    pub fn numeric_id(&self) -> i32 {
        self.id.to_i32()
    }

    pub fn temperature(&self) -> f32 {
        self.temperature
    }

    pub fn humidity(&self) -> f32 {
        self.humidity
    }

    pub fn resource_location(&self) -> &'static str {
        self.resource_location
    }
}

impl Default for Biome {
    fn default() -> Self {
        Self::from_id(1).unwrap()
    }
}

macro_rules! cast_biome_id_from_i32 {
    ($($t:ty),*) => {
        $(
            impl Into<$t> for Biome {
                fn into(self) -> $t {
                    self.numeric_id() as $t
                }
            }
        )*
    };
}

cast_biome_id_from_i32!(u8, i8, u16, i16, u32, i32, u64, i64);
